### PR TITLE
Bh/clone2

### PR DESCRIPTION
**Pull request by bheni**
This version of clone works on the table files directly.  It enumerates all the table files and downloads them.  It does not inspect the chunks as v1 did.

### CODE REVIEW

File: go/cmd/dolt/dolt.go [hunk 1]
```diff
@@ -83,42 +83,45 @@ func main() {
 func runMain() int {
 	args := os.Args[1:]
```
Comment (bheni) [inline go/cmd/dolt/dolt.go]: this isn't really part of this feature.  Just cleanup of debug flag processing.

File: go/utils/remotesrv/http.go [hunk 2]
```diff
@@ -130,6 +135,38 @@ func offsetAndLenFromRange(rngStr string) (int64, int64, error) {
...
+	if err != nil {
+		logger("failed to open file. file: " + path + " err: " + err.Error())
+		return http.StatusInternalServerError
+	}
+
+	n, err := io.Copy(writer, f)
+
+	if err != nil {
+		logger("failed to write data to response. err : " + err.Error())
+		return -1
```
Comment (zachmu) [inline go/utils/remotesrv/http.go]: Shouldn't this be an HTTP code?

File: go/utils/remotesrv/http.go [hunk 3]
```diff
@@ -130,6 +135,38 @@ func offsetAndLenFromRange(rngStr string) (int64, int64, error) {
...
+	}
+
+	if n != info.Size() {
+		logger(fmt.Sprintf("failed to write entire file to response. Copied %d of %d err: %v", n, info.Size(), err))
+		return -1
+	}
+
+	return -1
```
Comment (zachmu) [inline go/utils/remotesrv/http.go]: Should be http.ok?

File: go/cmd/dolt/commands/clone.go [hunk 4]
```diff
@@ -189,110 +187,68 @@ func createRemote(ctx context.Context, remoteName, remoteUrl string, params map[
 }
 
 func cloneRemote(ctx context.Context, srcDB *doltdb.DoltDB, remoteName, branch string, dEnv *env.DoltEnv) errhand.VerboseError {
-	var branches []ref.DoltRef
-	if len(branch) > 0 {
-		branches = []ref.DoltRef{ref.NewBranchRef(branch)}
-	} else {
-		var err error
-		branches, err = srcDB.GetBranches(ctx)
+	err := actions.Clone(ctx, srcDB, dEnv.DoltDB)
```
Comment (zachmu) [inline go/cmd/dolt/commands/clone.go]: Does this mean we lose progress indicators?

File: proto/dolt/services/remotesapi/v1alpha1/chunkstore.proto [hunk 5]
```diff
@@ -35,6 +35,8 @@ service ChunkStoreService {
   rpc Root(RootRequest) returns (RootResponse);
 
   rpc Commit(CommitRequest) returns (CommitResponse);
+
+  rpc EnumerateTables(EnumerateTablesRequest) returns (EnumerateTablesResponse);
```
[the remark this answers is not in the document]
Comment (bheni) [reply proto/dolt/services/remotesapi/v1alpha1/chunkstore.proto]: done

File: go/libraries/doltcore/remotestorage/retrying_cssclient.go [hunk 6]
```diff
@@ -123,3 +123,16 @@ func (c RetryingChunkStoreServiceClient) Commit(ctx context.Context, in *remotes
 
 	return resp, err
 }
+
+func (c RetryingChunkStoreServiceClient) EnumerateTables(ctx context.Context, in *remotesapi.EnumerateTablesRequest, opts ...grpc.CallOption) (*remotesapi.EnumerateTablesResponse, error) {
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/remotestorage/retrying_cssclient.go]: global replace :)

File: go/store/nbs/store.go [hunk 7]
```diff
@@ -823,3 +829,149 @@ func (nbs *NomsBlockStore) StatsSummary() string {
...
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.ModePerm)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return NomsBlockStoreTableSink{fileId, numChunks, f, nbs}, nil
+}
+
+// SetRootChunk sets the root chunk changes the root chunk hash from the previous value to the new root.
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/store/nbs/store.go]: fixed

File: go/libraries/doltcore/remotestorage/chunk_store.go [hunk 8]
```diff
@@ -772,3 +773,63 @@ func (dcs *DoltChunkStore) getDownloadWorkForLoc(ctx context.Context, getRange *
...
+
+	var tblFiles []nbs.TableFile
+	for _, nfo := range resp.TableInfo {
+		tblFiles = append(tblFiles, DoltRemoteTableFile{dcs, nfo})
+	}
+
+	return hash.New(resp.RootHash), tblFiles, nil
+}
+
+// SetRootChunk sets the root chunk changes the root chunk hash from the previous value to the new root.
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/remotestorage/chunk_store.go]: fixed

File: go/libraries/doltcore/remotestorage/chunk_store.go [hunk 9]
```diff
@@ -772,3 +773,63 @@ func (dcs *DoltChunkStore) getDownloadWorkForLoc(ctx context.Context, getRange *
...
+func (dcs *DoltChunkStore) SetRootChunk(ctx context.Context, root, previous hash.Hash) error {
+	panic("Not Implemented")
+}
+
+// DoltRemoteTableFile is an implementation of a TableFile that live in a DoltChunkStore
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/libraries/doltcore/remotestorage/chunk_store.go]: fixed

File: go/utils/remotesrv/http.go (hunk 2 above)
Comment (bheni) [reply go/utils/remotesrv/http.go]: fixed

File: go/cmd/dolt/commands/clone.go (hunk 4 above)
Comment (bheni) [reply go/cmd/dolt/commands/clone.go]: fixed

File: go/cmd/dolt/commands/clone.go [hunk 10]
```diff
@@ -189,110 +187,68 @@ func createRemote(ctx context.Context, remoteName, remoteUrl string, params map[
...
 		if err != nil {
-			return errhand.BuildDError("error: unable to find %v", branch).AddCause(err).Build()
+			return errhand.BuildDError("error: failed to list branches").AddCause(err).Build()
 		}
 
-		progChan := make(chan datas.PullProgress)
-		doneChan := make(chan struct{})
-		go progFunc(progChan, doneChan)
+		for _, brnch := range branches {
+			branch = brnch.GetPath()
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/cmd/dolt/commands/clone.go]: matches previous behavior

**Pull request merged**